File: writeups/83_Remove_Duplicates_from_Sorted_List/ruby/main.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  current_node = head
  while current_node
    if current_node.next
      if current_node.val == current_node.next.val
        current_node.next = current_node.next.next
        delete_duplicates(current_node)
      end
    end
    current_node = current_node.next
  end

  head
end

File: writeups/83_Remove_Duplicates_from_Sorted_List/typescript/main.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
  let currentNode: ListNode | null = head;
  while (currentNode) {
    if (currentNode.next) {
      if (currentNode.val === currentNode.next.val) {
        currentNode.next = currentNode.next.next;
        deleteDuplicates(currentNode);
      }
    }
    currentNode = currentNode.next;
  }

  return head;
}
